### DISCUSSION TITLE
Problem with server-only annotation when passing server component as props to client component

### DISCUSSION THREAD

**Discussion by ridhwaans** [vercel/next.js#67125] [Help] [answered]
### Summary

hi, I am following this pattern https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#supported-pattern-passing-server-components-to-client-components-as-props

I have the following setup
client component
```
interface ClientComponentNavProps {
  genres: Genre[];
  selectedLink: string;
  onGenreChange: (name: string) => void;
  children?: React.ReactNode;
}

export const ClientComponentNav = ({
  genres: initialGenres,
  selectedLink,
  onGenreChange,
  children,
}: ClientComponentNavProps) => {
return (
    <>
     <div>{stuff}</div>
     {children}
    </>
  );
}
```
page.tsx
```
'use client';
import { useState } from 'react'; //  need the use client because useState requires it
import { SliderComponent as ServerComponent } from '@/components/SliderComponent';
import { ClientComponentNav } from '@/components/ClientComponentNav';

const Page = () => {
    const [selectedLink, setSelectedLink] = useState('All');

  const handleGenreChange = (name: string) => {
    setSelectedLink(name);
  };
 return (
   <>
      <ClientComponentNav
              genres={queryData}
              selectedLink={selectedLink}
              onGenreChange={handleGenreChange}
            >
              <ServerCarousel genreName={selectedLink} />
            </GenresNav>
   </>
)
```
However, it fails to compile because the server component has an import `import { api } from '@/lib/trpc/server';` and `api` has the following
```
import 'server-only';

import { cache } from 'react';
import { headers } from 'next/headers';
...
export const api = ...
```
On the page, removing `use client` results in an error due to a `useState` dependency

It says that using 'server-only' is not allowed.
If I remove that it fails again with a header error
```
> 4 | import { headers } from 'next/headers';
You're importing a component that needs next/headers. That only works in a Server Component which is not supported in the pages/ directory.
```
My workaround is recreating several components as client components. 
 How do i resolve this without a ton of rewriting? 
thanks

full error
```
./apps/web/src/lib/trpc/server.ts:1:1
Ecmascript file had an error
> 1 | import 'server-only';
    | ^^^^^^^^^^^^^^^^^^^^^
  2 |
  3 | import { cache } from 'react';
  4 | import { headers } from 'next/headers';

You're importing a component that needs server-only. That only works in a Server Component which is not supported in the pages/ directory. Read more: https://nextjs.org/docs/getting-started/react-essentials#server-components
```

### Additional information

```js
"next": "14.2.3". 
using a monorepo
```

**edit 1**: could be a bug https://github.com/dubinc/dub/issues/258
https://stackoverflow.com/questions/77401047/server-component-imported-into-a-this-doesnt-work-client-component-and-wrapped

**edit 2**: I think i need to just use server actions

**Comment by icyJoseph** [comment, marked as answer]
There's a concept issue here, you can't really have a stateful component own a server component. The moment state changes, things will break. You did well composition with `ClientComponentNav`, but did not when building the `Page` component.

In other words, you are still doing the Unsupported Pattern here (also shown in the link you posted).

To quickly get this working, I recommend, moving the useState hook into the `ClientComponentNav`, remove `use client` from Page, and have it set a url query, which the `ServerCarousel`, can read from its page parent. You'll need to sort out a bunch of other details though.

**Reply by ridhwaans** [reply]
Yes, the problem was page is a client component when it should be a RSC
I started using searchparams to make server actions
```javascript
"use client";
import { usePathname, useRouter, useSearchParams } from "next/navigation";
import { type ChangeEvent, useTransition } from "react";

interface ClientComponentNavProps {
  genres: Genre[];
  selectedLink: string;
  onGenreChange: (name: string) => void;
  children?: React.ReactNode;
}

export const ClientComponentNav = ({
  genres: initialGenres,
  selectedLink,
  onGenreChange,
  children,
}: ClientComponentNavProps) => {
const searchParams = useSearchParams();
const [isPending, startTransition] = useTransition();
  
  const handleGenreChange = (name) => {
    const updatedSearchParams = new URLSearchParams(searchParams);

    if (name) {
      updatedSearchParams.set("q", name);
    } else {
      updatedSearchParams.delete("q");
    }

    startTransition(() => {
      router.replace(`${pathname}?${updatedSearchParams.toString()}`);
    });
  };
  
return (
    <>
     <div onClick={() => handleGenreChange(genre.name as string)}>{stuff}</div>
     {children}
    </>
  );
  
  
export default async function Page({ searchParams }: { searchParams: { q?: string } }) {
...
}
```

however, it deemed to be too slow as server side re-renders were taking >= ~2 seconds

I changed `ServerCarousel` into a `ClientCarousel` and moved it into `ClientComponentNav` and use state there
Page is now a server component and it works 
Thank you
